GitHub Deployments: Restore focus after selecting GitHub account

epeicher opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Closes DOTDASH-642

Further context: pdtkmj-42o-p2#comment-7772

## Proposed Changes

* Adds an effect that restores the focus to the GitHub Account dropdown when it changes
* Adds a piece of state that stores if the GH Account is changing

> [!NOTE]
> ~~This seems to complicate a bit the code to solve a use case that is not that bad. The loop is just three fields and then the user can continue tabbing. Happy to hear others view on this.~~
Eventually we decided to move forward with these changes as it improves accessibility and the changes are not that complex. Please see comments on the PR below

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* There is a loop in the Tab order when changing the GH account

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply this branch and run Calypso locally or open the Calypso Live link below
* Navigate to `/v2/sites/:siteSlug/settings/repositories` and click on `Connect repository` or use an existing repository and click on `Configure repository` on the three dots menu
* Use the Tab key to navigate through the fields, and when navigating to the GitHub Account selector, change the value to something else
* Check you can continue selecting the next field by using the Tab key, and the selection does not start over

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

epeicher committed:
GitHub Deployments: Restore focus after selecting GitHub account

epeicher committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into dotdash-642-fix-tab-order-on-the-connect-repository-form

ivan-ottinger reviewed (approved):
Thanks for working on this fix, Roberto. The changes look good and work as expected:

https://github.com/user-attachments/assets/fe52526d-5fe4-4a60-a911-0bedd1b24956

> This seems to complicate a bit the code to solve a use case that is not that bad. The loop is just three fields and then the user can continue tabbing. Happy to hear others view on this.

Yeah, that's a good point. I looked into other fixes, but no better solution came to my mind.

I think it would also be ok to leave things as they are - as I don't consider this issue as something significant. Especially when considering the extra code needed to address it.

gcsecsey reviewed (commented):
Thanks @epeicher for tackling this issue! 🙌 This is working great, and I think the solution is minimal given the constraints.

https://github.com/user-attachments/assets/6e996115-4519-49c2-b9b9-c990282d50e9

> This seems to complicate a bit the code to solve a use case that is not that bad. The loop is just three fields and then the user can continue tabbing. Happy to hear others view on this.

I agree that we might be chasing just a slight improvement here. 

However, the focus reset seems to be the side-effect of rendering dependent fields, eg. when choosing the GitHub account, we update the list of repositories. This same thing happens, if we select the repository:

https://github.com/user-attachments/assets/8cb38c6b-af70-4a86-aa56-95de09046afa

So if we're keeping this solution, I think we should also extend it to the repositories selector.

@StevenDufresne what do you think, is it worth the extra code to fix the tab order? Thanks!

epeicher commented:
Thanks @p-jackson for the input! I have been profiling a bit this component and I see that the re-render of the form is due to multiple causes, plase see that there are 6 hooks that cause the form to re-render:
<img width="305" height="218" alt="CleanShot 2025-10-20 at 13 22 46@2x" src="https://github.com/user-attachments/assets/f5361102-f93a-41c9-aa4b-3dd3d36e6517" />

Apart from the re-rendering of [the specific control](https://github.com/Automattic/wp-calypso/pull/106521/files#diff-d82db2f79917ed6d32200b1317b3c96843fe507ec6fe8b04d8b9b6f33567bebeR562) that also looses the focus. 

So I think we can solve the focus issue as part of this PR and work on minimizing the re-renders as part of a follow-up. If we are able to elmiinate the re-renders, we can remove the focus-specific code.

> The ability to tab through form fields is required for web accessibility. We need to support that.

@StevenDufresne , got it, I will add the code to keep the focus on the DropDown that looses it, I think it only applies to the one you discovered and the next one.

> we're [intentionally re-rendering](https://github.com/Automattic/wp-calypso/blob/26ffd9b172dfe397a60f61adcad3db035806e847/client/dashboard/sites/settings-repositories/connect-repository-form.tsx#L627) to fix some validation issues. I would question whether unmounting/mounting the component in its entirety is the best behavior.

Well spotted! That was one of my first tries, but the issue still happens, there are a lot of queries to retrieve dependent dropdown list of values that cause the re-renders, so that could be tackled as part of a follow-up. This PR can fix the immediate issue.

epeicher committed:
Simplify code

p-jackson reviewed (approved):
> So I think we can solve the focus issue as part of this PR and work on minimizing the re-renders as part of a follow-up. If we are able to elmiinate the re-renders, we can remove the focus-specific code.

Yes I think this is the right approach. DOM stuff like focus should "just work" when all the React renders are clean. But we also can't leave the user experience broken. It is good to know that we can clean up the manual focusing code if we are able to tidy up the rendering somehow.

I've tested and it is working as expected. The same issue is still on the repository selector, but that can be a different PR.

I also personally tried doing the fix taking an imperative approach too.

https://github.com/Automattic/wp-calypso/commit/5d5d493a5d6e6e4b65bfd8bf4231050af92d5b77

When something is inherently imperative (like setting focus) I find it can be clearer to drive the element directly rather than depending on props. For example, it is a little bit subtle the way the fix relies on the `shouldRestoreFocus` prop being set to `false` so that it is reset and ready to go for the next time `shouldRestoreFocus` gets set to `true` again.

You don't have to take this approach though. I just wanted to see if this would work too 😄

epeicher committed:
Use useImperativeHandle to restore the focus

This requires a setTimeout as the component is destroyed multiple times

epeicher committed:
Revert "Use useImperativeHandle to restore the focus"

This reverts commit dbd2393dd0ec4b9a70b79be98a8fe25a01a4e2d1.

epeicher committed:
Use useEffect with a flag to restore the focus

This option keeps the flag to restore the focus active during the multiple re-renders

epeicher commented on client/dashboard/sites/settings-repositories/connect-repository-form.tsx:
```diff
@@ -110,6 +123,7 @@ const RepositorySelector = ( {
 				__next40pxDefaultSize
 				__nextHasNoMarginBottom
 				allowReset
+				expandOnFocus={ false }
```
Required to prevent the expansion of the dropdown when getting the focus back. The combobox still expands the options when selected with the mouse or clicking on the arrows, so it keeps the keyboard functionality

epeicher committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into dotdash-642-fix-tab-order-on-the-connect-repository-form

epeicher commented:
Thanks @p-jackson for your suggestion! I have been iterating over [that approach](https://github.com/Automattic/wp-calypso/commit/5d5d493a5d6e6e4b65bfd8bf4231050af92d5b77) and also trying to expose a `getTheFocus` function [through useImperativeHandle](https://react.dev/reference/react/useImperativeHandle) to the parent component, but the multiple re-renders have made it tricky to provide the focus to the proper `input` component on the `RepositorySelector`. 

Eventually, I have decided to go through the [flag-based approach](https://github.com/Automattic/wp-calypso/pull/106521/files#diff-d82db2f79917ed6d32200b1317b3c96843fe507ec6fe8b04d8b9b6f33567bebeR335) combined with the `useEffect` in the component receiving the focus and that seems to be more reliable.

I'll proceed with this for now if it passes the reviews but I am happy to study better alternatives in other iterations.

@gcsecsey, @ivan-ottinger, could you please try another review if possible? 🙏  With the latest changes all the dropdowns on the form should keep the tab order after selecting values.

ivan-ottinger commented on client/dashboard/sites/settings-repositories/connect-repository-form.tsx:
```diff
@@ -92,9 +92,22 @@ const RepositorySelector = ( {
+				input.focus();
+			}
+		}
+	}, [ shouldRestoreFocus, currentValue, field.elements ] );
```
The proposed code above works as expected. I am just worried how fragile is using class selector there.

What if we used `useRef` for the `RepositorySelector` as well? We could consider enclosing `<ComboboxControl>` inside a `<div ref={ comboboxRef }>` and then have the `useEffect` part like this:

```jsx
	const comboboxRef = useRef< HTMLInputElement | null >( null );

	useEffect( () => {
		if ( shouldRestoreFocus && comboboxRef.current ) {
			const input =
				comboboxRef.current.querySelector< HTMLInputElement >( 'input[role="combobox"]' );
			if ( input ) {
				input.focus();
			}
		}
	}, [ shouldRestoreFocus ] );
```

But, you have been looking into this issue for much longer, so I might be missing something.

epeicher replied:
Thanks for the suggestion, I think is a better approach! I used the class selector because the ComboboxControl does not provide a ref, but I think in this scenario, the div wrapper improves the code. Also, the selector based on role avoid looking for the label, so applying that too 🙌

epeicher replied:
[2163953](https://github.com/Automattic/wp-calypso/pull/106521/commits/21639533f65cca12ad0c175056ae85b897a37fa3)

ivan-ottinger commented on client/dashboard/sites/settings-repositories/connect-repository-form.tsx:
```diff
@@ -145,15 +159,24 @@ const RepositorySelector = ( {
+		if ( selectRef.current && shouldRestoreFocus ) {
+			selectRef.current.focus();
+		}
+	}, [ selectRef, shouldRestoreFocus ] );
```
I think we could omit `selectRef` here.

```suggestion
	}, [ shouldRestoreFocus ] );
```

epeicher replied:
That's right, well spotted! Removing it

epeicher replied:
[2163953](https://github.com/Automattic/wp-calypso/pull/106521/commits/21639533f65cca12ad0c175056ae85b897a37fa3)

ivan-ottinger commented on client/dashboard/sites/settings-repositories/connect-repository-form.tsx:
```diff
@@ -306,9 +332,25 @@ export const ConnectRepositoryForm = ( {
+
+			if (
+				'selectedRepositoryId' in updates &&
+				updates.selectedRepositoryId !== prev.selectedRepositoryId
```
When we keep the `updates.selectedRepositoryId !== prev.selectedRepositoryId` check in and the user selects the same repository again, we lose focus.


https://github.com/user-attachments/assets/b47aaa37-d41a-41cf-9953-9020a66d7c1b

Because of that, we could remove that check:

```suggestion
				'selectedRepositoryId' in updates
```

epeicher replied:
This was a great catch, thanks for that! I have reproduced it locally, applying your suggestion 👍

epeicher replied:
[2163953](https://github.com/Automattic/wp-calypso/pull/106521/commits/21639533f65cca12ad0c175056ae85b897a37fa3)

ivan-ottinger reviewed (commented):
Thank you for exploring the options, Roberto!

The changes work as expected. I only observed one issue when selecting the same repository again. I also shared two more thoughts on alternative changes. Please see the inline comments below.

The core focus functionality works nicely as can be seen in the recording below. 🙂 

https://github.com/user-attachments/assets/efae4421-0dbb-4f50-be0b-e2ae131bdf05

epeicher committed:
Applying code review suggestions

epeicher committed:
Merge branch 'trunk' of github.com:Automattic/wp-calypso into dotdash-642-fix-tab-order-on-the-connect-repository-form

ivan-ottinger commented on client/dashboard/sites/settings-repositories/connect-repository-form.tsx:
```diff
@@ -92,9 +92,21 @@ const RepositorySelector = ( {
+				input.focus();
+			}
+		}
+	}, [ shouldRestoreFocus, currentValue, field.elements ] );
```
I think it should be enough to keep `shouldRestoreFocus` only:

```suggestion
	}, [ shouldRestoreFocus ] );
```

ivan-ottinger reviewed (approved):
Thank you for the changes, Roberto!

All is working well now. 👌🏼 I only left one small suggestion below. Apart from that, I did not spot any issues. 🙂

epeicher committed:
Remove leftovers

epeicher committed:
Merge branches 'dotdash-642-fix-tab-order-on-the-connect-repository-form' and 'trunk' of github.com:Automattic/wp-calypso into dotdash-642-fix-tab-order-on-the-connect-repository-form
